File: crates/cli/plugins/go.lua
-- ponytail: control-char detection is ASCII 0x00-0x1F + DEL only, not full
-- Unicode C1 range (0x80-0x9F) like Rust's char::is_control(); upgrade if a
-- plugin ever needs to key on non-ASCII control characters.
local function is_control_byte(b)
  return b < 0x20 or b == 0x7f
end

local function json_escape(str)
  local out = { '"' }
  for i = 1, #str do
    local b = str:byte(i)
    local ch = str:sub(i, i)
    if ch == '"' then
      out[#out + 1] = '\\"'
    elseif ch == '\\' then
      out[#out + 1] = '\\\\'
    elseif b == 8 then
      out[#out + 1] = '\\b'
    elseif b == 9 then
      out[#out + 1] = '\\t'
    elseif b == 10 then
      out[#out + 1] = '\\n'
    elseif b == 12 then
      out[#out + 1] = '\\f'
    elseif b == 13 then
      out[#out + 1] = '\\r'
    elseif b < 32 then
      out[#out + 1] = string.format('\\u%04x', b)
    else
      out[#out + 1] = ch
    end
  end
  out[#out + 1] = '"'
  return table.concat(out)
end

local function contains_any_byte(str, targets)
  for i = 1, #str do
    local b = str:byte(i)
    for _, target in ipairs(targets) do
      if b == target then
        return true
      end
    end
  end
  return false
end

local function has_control_char(str)
  for i = 1, #str do
    if is_control_byte(str:byte(i)) then
      return true
    end
  end
  return false
end

local function needs_custom_json_key(name)
  return name == "-" or name:find(",", 1, true) ~= nil or has_control_char(name)
end

local function needs_interpreted_tag_literal(name)
  return contains_any_byte(name, { 0x60, 0x5c, 0x22 }) or has_control_char(name)
end

local function render_json_tag(name)
  local tag = "json:" .. json_escape(name)
  if needs_interpreted_tag_literal(name) then
    return json_escape(tag)
  end
  return "`" .. tag .. "`"
end

local function render_struct_tag(name)
  if needs_custom_json_key(name) then
    return '`json:"-"`'
  end
  return render_json_tag(name)
end

local function is_go_ident_start(ch)
  return ch == "_" or ch:match("%a") ~= nil
end

local function is_go_ident_char(ch)
  return ch == "_" or ch:match("%w") ~= nil
end

local function render_type_name(name)
  local out = {}
  for i = 1, #name do
    local ch = name:sub(i, i)
    -- NOTE: not `(i == 1) and is_go_ident_start(ch) or is_go_ident_char(ch)`
    -- -- that classic Lua ternary idiom silently breaks when
    -- is_go_ident_start(ch) is false, falling through to
    -- is_go_ident_char(ch) instead of staying false.
    local valid
    if i == 1 then
      valid = is_go_ident_start(ch)
    else
      valid = is_go_ident_char(ch)
    end
    if valid then
      out[#out + 1] = ch
    elseif i == 1 then
      out[#out + 1] = "_"
      if ch:match("%d") then
        out[#out + 1] = ch
      end
    elseif out[#out] ~= "_" then
      out[#out + 1] = "_"
    end
  end
  local result = table.concat(out)
  return result == "" and "Root" or result
end

local function render_field_name(name)
  local out = {}
  local part = {}
  local function flush()
    if #part == 0 then
      return
    end
    local first = part[1]
    if #out == 0 and first:match("%d") then
      out[#out + 1] = "X"
    end
    out[#out + 1] = first:upper()
    for i = 2, #part do
      out[#out + 1] = part[i]:lower()
    end
    part = {}
  end
  for i = 1, #name do
    local ch = name:sub(i, i)
    if ch:match("%w") then
      part[#part + 1] = ch
    else
      flush()
    end
  end
  flush()
  local result = table.concat(out)
  return result == "" and "Field" or result
end

local function render_field_names(fields)
  local used = {}
  local names = {}
  for _, field in ipairs(fields) do
    local base = render_field_name(field.name)
    used[base] = (used[base] or 0) + 1
    local n = used[base]
    names[#names + 1] = (n == 1) and base or (base .. tostring(n))
  end
  return names
end

local function render_type(ty)
  local kind = ty.kind
  if kind == "any" then
    return "interface{}"
  elseif kind == "bool" then
    return "bool"
  elseif kind == "integer" then
    return "int64"
  elseif kind == "unsigned_integer" then
    return "uint64"
  elseif kind == "float" then
    return "float64"
  elseif kind == "string" then
    return "string"
  elseif kind == "named" then
    return render_type_name(ty.name)
  elseif kind == "array" then
    return "[]" .. render_type(ty.item)
  end
  error("unknown type kind: " .. tostring(kind))
end

local function render_field_type(field)
  local ty = render_type(field.ty)
  if field.optional then
    return "*" .. ty
  end
  return ty
end

local function named_type_needs_custom_json(named)
  for _, field in ipairs(named.fields) do
    if needs_custom_json_key(field.name) then
      return true
    end
  end
  return false
end

local function render_custom_json_methods(named, field_names, out)
  local type_name = render_type_name(named.name)

  out[#out + 1] = "func (value *" .. type_name .. ") UnmarshalJSON(data []byte) error {\n"
  out[#out + 1] = "\ttype plain " .. type_name .. "\n"
  out[#out + 1] = "\traw := map[string]json.RawMessage{}\n"
  out[#out + 1] = "\tif err := json.Unmarshal(data, &raw); err != nil {\n"
  out[#out + 1] = "\t\treturn err\n"
  out[#out + 1] = "\t}\n"

  for i, field in ipairs(named.fields) do
    if needs_custom_json_key(field.name) then
      local field_name = field_names[i]
      out[#out + 1] = "\tif field, ok := raw[" .. json_escape(field.name) .. "]; ok {\n"
      out[#out + 1] = "\t\tif err := json.Unmarshal(field, &value." .. field_name .. "); err != nil {\n"
      out[#out + 1] = "\t\t\treturn err\n"
      out[#out + 1] = "\t\t}\n"
      out[#out + 1] = "\t}\n"
      out[#out + 1] = "\tdelete(raw, " .. json_escape(field.name) .. ")\n"
    end
  end

  out[#out + 1] = "\trest, err := json.Marshal(raw)\n"
  out[#out + 1] = "\tif err != nil {\n"
  out[#out + 1] = "\t\treturn err\n"
  out[#out + 1] = "\t}\n"
  out[#out + 1] = "\treturn json.Unmarshal(rest, (*plain)(value))\n"
  out[#out + 1] = "}\n\n"

  out[#out + 1] = "func (value " .. type_name .. ") MarshalJSON() ([]byte, error) {\n"
  out[#out + 1] = "\ttype plain " .. type_name .. "\n"
  out[#out + 1] = "\tencoded, err := json.Marshal(plain(value))\n"
  out[#out + 1] = "\tif err != nil {\n"
  out[#out + 1] = "\t\treturn nil, err\n"
  out[#out + 1] = "\t}\n"
  out[#out + 1] = "\traw := map[string]json.RawMessage{}\n"
  out[#out + 1] = "\tif err := json.Unmarshal(encoded, &raw); err != nil {\n"
  out[#out + 1] = "\t\treturn nil, err\n"
  out[#out + 1] = "\t}\n"
  out[#out + 1] = "\tvar payload []byte\n"

  for i, field in ipairs(named.fields) do
    if needs_custom_json_key(field.name) then
      local field_name = field_names[i]
      out[#out + 1] = "\tpayload, err = json.Marshal(value." .. field_name .. ")\n"
      out[#out + 1] = "\tif err != nil {\n"
      out[#out + 1] = "\t\treturn nil, err\n"
      out[#out + 1] = "\t}\n"
      out[#out + 1] = "\traw[" .. json_escape(field.name) .. "] = payload\n"
    end
  end

  out[#out + 1] = "\treturn json.Marshal(raw)\n"
  out[#out + 1] = "}\n\n"
end

local function render_named_type(named, out)
  local field_names = render_field_names(named.fields)

  out[#out + 1] = "type " .. render_type_name(named.name) .. " struct {\n"
  for i, field in ipairs(named.fields) do
    out[#out + 1] = "\t"
      .. field_names[i]
      .. " "
      .. render_field_type(field)
      .. " "
      .. render_struct_tag(field.name)
      .. "\n"
  end
  out[#out + 1] = "}\n\n"

  if named_type_needs_custom_json(named) then
    render_custom_json_methods(named, field_names, out)
  end
end

return {
  key = "go",
  render = function(document)
    local needs_json_import = false
    for _, named in ipairs(document.types) do
      if named_type_needs_custom_json(named) then
        needs_json_import = true
        break
      end
    end

    local out = { "package models\n\n" }
    if needs_json_import then
      out[#out + 1] = 'import "encoding/json"\n\n'
    end

    for _, named in ipairs(document.types) do
      render_named_type(named, out)
    end

    if document.root.kind ~= "named" then
      out[#out + 1] = "type " .. render_type_name(document.root_name) .. " = " .. render_type(document.root) .. "\n"
    end

    return table.concat(out)
  end,
}

File: crates/core/src/render/go.rs
//! Go renderer.
//!
//! Native port of `crates/cli/plugins/go.lua`. The two implementations must
//! stay in sync byte for byte — `crates/cli/tests/render_parity.rs` enforces
//! that, so any change here needs the matching change in the plugin (and vice
//! versa).

use super::{deduplicate, json_escape, sanitize_type_name};
use crate::schema::{Document, Field, NamedType, TypeExpr};

/// Renders `document` as a Go `package models` source file.
///
/// Emits one `struct` per named type, plus `UnmarshalJSON`/`MarshalJSON`
/// methods for the structs holding at least one JSON key a struct tag cannot
/// express, then — only when the root is not itself a named type — a trailing
/// type alias for the root.
pub fn render(document: &Document) -> String {
    let mut out = String::from("package models\n\n");

    if document.types.iter().any(named_type_needs_custom_json) {
        out.push_str("import \"encoding/json\"\n\n");
    }

    for named in &document.types {
        render_named_type(named, &mut out);
    }

    if !matches!(document.root, TypeExpr::Named { .. }) {
        out.push_str("type ");
        out.push_str(&render_type_name(&document.root_name));
        out.push_str(" = ");
        out.push_str(&render_type(&document.root));
        out.push('\n');
    }

    out
}

fn render_named_type(named: &NamedType, out: &mut String) {
    let field_names = render_field_names(&named.fields);

    out.push_str("type ");
    out.push_str(&render_type_name(&named.name));
    out.push_str(" struct {\n");

    for (field, field_name) in named.fields.iter().zip(&field_names) {
        out.push('\t');
        out.push_str(field_name);
        out.push(' ');
        out.push_str(&render_field_type(field));
        out.push(' ');
        out.push_str(&render_struct_tag(&field.name));
        out.push('\n');
    }

    out.push_str("}\n\n");

    if named_type_needs_custom_json(named) {
        render_custom_json_methods(named, &field_names, out);
    }
}

/// Emits hand-rolled marshal/unmarshal methods for the keys `encoding/json`
/// cannot round-trip through a struct tag.
///
/// Only the offending fields get explicit handling; everything else still goes
/// through the generated `plain` alias so the ordinary struct tags keep working.
fn render_custom_json_methods(named: &NamedType, field_names: &[String], out: &mut String) {
    let type_name = render_type_name(&named.name);

    out.push_str("func (value *");
    out.push_str(&type_name);
    out.push_str(") UnmarshalJSON(data []byte) error {\n");
    out.push_str("\ttype plain ");
    out.push_str(&type_name);
    out.push('\n');
    out.push_str("\traw := map[string]json.RawMessage{}\n");
    out.push_str("\tif err := json.Unmarshal(data, &raw); err != nil {\n");
    out.push_str("\t\treturn err\n");
    out.push_str("\t}\n");

    for (field, field_name) in named.fields.iter().zip(field_names) {
        if !needs_custom_json_key(&field.name) {
            continue;
        }

        let key = json_escape(&field.name);

        out.push_str("\tif field, ok := raw[");
        out.push_str(&key);
        out.push_str("]; ok {\n");
        out.push_str("\t\tif err := json.Unmarshal(field, &value.");
        out.push_str(field_name);
        out.push_str("); err != nil {\n");
        out.push_str("\t\t\treturn err\n");
        out.push_str("\t\t}\n");
        out.push_str("\t}\n");
        out.push_str("\tdelete(raw, ");
        out.push_str(&key);
        out.push_str(")\n");
    }

    out.push_str("\trest, err := json.Marshal(raw)\n");
    out.push_str("\tif err != nil {\n");
    out.push_str("\t\treturn err\n");
    out.push_str("\t}\n");
    out.push_str("\treturn json.Unmarshal(rest, (*plain)(value))\n");
    out.push_str("}\n\n");

    out.push_str("func (value ");
    out.push_str(&type_name);
    out.push_str(") MarshalJSON() ([]byte, error) {\n");
    out.push_str("\ttype plain ");
    out.push_str(&type_name);
    out.push('\n');
    out.push_str("\tencoded, err := json.Marshal(plain(value))\n");
    out.push_str("\tif err != nil {\n");
    out.push_str("\t\treturn nil, err\n");
    out.push_str("\t}\n");
    out.push_str("\traw := map[string]json.RawMessage{}\n");
    out.push_str("\tif err := json.Unmarshal(encoded, &raw); err != nil {\n");
    out.push_str("\t\treturn nil, err\n");
    out.push_str("\t}\n");
    out.push_str("\tvar payload []byte\n");

    for (field, field_name) in named.fields.iter().zip(field_names) {
        if !needs_custom_json_key(&field.name) {
            continue;
        }

        out.push_str("\tpayload, err = json.Marshal(value.");
        out.push_str(field_name);
        out.push_str(")\n");
        out.push_str("\tif err != nil {\n");
        out.push_str("\t\treturn nil, err\n");
        out.push_str("\t}\n");
        out.push_str("\traw[");
        out.push_str(&json_escape(&field.name));
        out.push_str("] = payload\n");
    }

    out.push_str("\treturn json.Marshal(raw)\n");
    out.push_str("}\n\n");
}

fn named_type_needs_custom_json(named: &NamedType) -> bool {
    named
        .fields
        .iter()
        .any(|field| needs_custom_json_key(&field.name))
}

fn render_type(ty: &TypeExpr) -> String {
    match ty {
        TypeExpr::Any => "interface{}".to_string(),
        TypeExpr::Bool => "bool".to_string(),
        TypeExpr::Integer => "int64".to_string(),
        TypeExpr::UnsignedInteger => "uint64".to_string(),
        TypeExpr::Float => "float64".to_string(),
        TypeExpr::String => "string".to_string(),
        TypeExpr::Named { name } => render_type_name(name),
        TypeExpr::Array { item } => format!("[]{}", render_type(item)),
    }
}

/// Optional fields become pointers so a missing key stays distinguishable from
/// the zero value.
fn render_field_type(field: &Field) -> String {
    let ty = render_type(&field.ty);

    if field.optional {
        return format!("*{ty}");
    }

    ty
}

/// `$` is not an identifier character in Go, unlike in TypeScript.
fn render_type_name(name: &str) -> String {
    sanitize_type_name(name, false)
}

/// Assigns every field an exported Go name, disambiguating collisions.
///
/// Repeats get a bare counter suffix (`Name`, `Name2`, `Name3`) — no
/// separator — because Go field names are camel case, not snake case.
fn render_field_names(fields: &[Field]) -> Vec<String> {
    let bases: Vec<String> = fields
        .iter()
        .map(|field| render_field_name(&field.name))
        .collect();

    deduplicate(&bases, "")
}

/// Turns a JSON key into an exported Go field name.
///
/// Splits on every non-alphanumeric byte, then upper-cases the first byte of
/// each run and lower-cases the rest — so `userID` becomes `Userid`, not
/// `UserID`. A run of digits at the very start is prefixed with `X` to keep the
/// identifier legal, and a name with no alphanumerics at all falls back to
/// `Field`.
///
/// Deliberately byte-wise and ASCII-only: Lua's `%w` class and `string.upper`
/// only cover ASCII under the C locale, so every byte of a multi-byte UTF-8
/// sequence acts as a separator. Iterating `chars()` here would keep letters the
/// plugin drops, breaking parity.
fn render_field_name(name: &str) -> String {
    let mut out = String::with_capacity(name.len() + 1);
    let mut part: Vec<u8> = Vec::new();

    for byte in name.bytes() {
        if byte.is_ascii_alphanumeric() {
            part.push(byte);
        } else {
            flush_field_name_part(&mut part, &mut out);
        }
    }

    flush_field_name_part(&mut part, &mut out);

    if out.is_empty() {
        return "Field".to_string();
    }

    out
}

/// Appends the pending run of alphanumerics to `out`, then clears it.
fn flush_field_name_part(part: &mut Vec<u8>, out: &mut String) {
    let Some(&first) = part.first() else {
        return;
    };

    // Only the leading run can start the identifier, so only it needs the
    // digit guard.
    if out.is_empty() && first.is_ascii_digit() {
        out.push('X');
    }

    // Every byte in `part` is ASCII, so the `as char` casts are lossless.
    out.push(first.to_ascii_uppercase() as char);
    for byte in &part[1..] {
        out.push(byte.to_ascii_lowercase() as char);
    }

    part.clear();
}

/// Reports whether `name` needs the generated `UnmarshalJSON`/`MarshalJSON`
/// methods rather than a plain struct tag.
///
/// `encoding/json` reads `-` as "skip this field" and treats a comma as the
/// start of tag options, so neither can appear in a tag as a literal key.
/// Control characters are excluded too, since they cannot survive a Go raw
/// string literal. Distinct from [`needs_interpreted_tag_literal`], which only
/// decides how to *quote* a key the tag can still carry.
fn needs_custom_json_key(name: &str) -> bool {
    name == "-" || name.contains(',') || has_control_char(name)
}

/// Reports whether `name`'s struct tag has to be an interpreted (double-quoted)
/// Go string literal instead of a raw (backtick) one.
///
/// A raw literal cannot contain a backtick, and the tag text itself already
/// contains the JSON-escaped key, so a backslash or double quote in `name`
/// would also have to be spelled out. Distinct from [`needs_custom_json_key`]:
/// these keys still round-trip through the tag, they just need heavier quoting.
fn needs_interpreted_tag_literal(name: &str) -> bool {
    name.bytes().any(|byte| matches!(byte, b'`' | b'\\' | b'"')) || has_control_char(name)
}

/// Renders the `json:"..."` struct tag for `name`, quoting as lightly as it can.
///
/// Falls back to `json:"-"` for keys [`needs_custom_json_key`] rejects; those
/// are carried by the generated marshal methods instead.
fn render_struct_tag(name: &str) -> String {
    if needs_custom_json_key(name) {
        return "`json:\"-\"`".to_string();
    }

    render_json_tag(name)
}

fn render_json_tag(name: &str) -> String {
    let tag = format!("json:{}", json_escape(name));

    if needs_interpreted_tag_literal(name) {
        return json_escape(&tag);
    }

    format!("`{tag}`")
}

fn has_control_char(value: &str) -> bool {
    value.bytes().any(is_control_byte)
}

// ponytail: control-char detection is ASCII 0x00-0x1F + DEL only, not full
// Unicode C1 range (0x80-0x9F) like Rust's char::is_control(); upgrade if a
// plugin ever needs to key on non-ASCII control characters.
fn is_control_byte(byte: u8) -> bool {
    byte < 0x20 || byte == 0x7f
}
